node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java

```diff
@@ -34,76 +31,71 @@
 public class InfrastructureProvisioner extends Maintainer {
 
     private static final Logger logger = Logger.getLogger(InfrastructureProvisioner.class.getName());
-    private static final List<HostedVespaApplication> HOSTED_VESPA_APPLICATIONS = ImmutableList.of(
-            ConfigServerApplication.CONFIG_SERVER_APPLICATION,
-            ConfigServerHostApplication.CONFIG_SERVER_HOST_APPLICATION,
-            ProxyHostApplication.PROXY_HOST_APPLICATION,
-            ControllerApplication.CONTROLLER_APPLICATION,
-            ControllerHostApplication.CONTROLLER_HOST_APPLICATION);
 
     private final Provisioner provisioner;
     private final InfrastructureVersions infrastructureVersions;
+    private final DuperModelInfraApi duperModel;
 
     public InfrastructureProvisioner(Provisioner provisioner, NodeRepository nodeRepository,
-                                     InfrastructureVersions infrastructureVersions, Duration interval, JobControl jobControl) {
+                                     InfrastructureVersions infrastructureVersions, Duration interval, JobControl jobControl,
+                                     DuperModelInfraApi duperModel) {
         super(nodeRepository, interval, jobControl);
         this.provisioner = provisioner;
         this.infrastructureVersions = infrastructureVersions;
+        this.duperModel = duperModel;
     }
 
     @Override
     protected void maintain() {
-        for (HostedVespaApplication application: HOSTED_VESPA_APPLICATIONS) {
+        for (InfraApplicationApi application: duperModel.getSupportedInfraApplications()) {
             try (Mutex lock = nodeRepository().lock(application.getApplicationId())) {
-                Optional<Version> version = getTargetVersion(application.getCapacity().type());
-                if (! version.isPresent()) continue;
+                NodeType nodeType = application.getCapacity().type();
+
+                Optional<Version> targetVersion = infrastructureVersions.getTargetVersionFor(nodeType);
+                if (!targetVersion.isPresent()) {
+                    logger.log(LogLevel.DEBUG, "Skipping provision of " + nodeType + ": No target version set");
+                    duperModel.infraApplicationRemoved(application.getApplicationId());
```

Comment:
It looks like its not possible to unset a target version once it has been set, so in that regard this is fine... 

However, if that were to change, this should also actually deallocate nodes allocated to this application? As this stand right now, those nodes would still be in `active`, but `DuperModel` wouldn't know anything about them?

Reply:
Yes, I think the InfrastructureProvisioner should then deallocate those nodes, and it sounds right that the DuperModel would not know about that application then.

Reply:
Right, it should, but it doesn't. Remove the removal and add a TODO instead?